### PR TITLE
Make the `/facet-search` route use the `sortFacetValuesBy` setting

### LINKED ISSUE #4423: Make `sortFacetValuesBy` work with `/facet-search` endpoint
Related product team resources: [PRD](https://www.notion.so/meilisearch/Faceted-search-improvements-7c4387305e4540a29d2f31ef6720dc7d) (_internal only_)
Related spec: WIP

## Motivation

Bring consistency in our API, so removes confusion. Also, requested by users.

## Usage

`sortFacetValuesBy` only works on the `/search` endpoint. Should work for `/facet-search` too.

Note for the @meilisearch/docs-team: Using the new `sortFacetValuesBy` setting can slow down the `/facet-search` requests as Meilisearch iterates over the whole list of facet values and computes the count of documents on every entry. That is hardly or even impossible to optimize correctly.

## TODO

- [ ] Merge changes into `main`
- [ ] Update the spec

## Impacted teams

@meilisearch/docs-team

### PR DESCRIPTION
**Pull request by Kerollmops**
This PR fixes #4423 by ensuring that the `/facet-search` route uses the `sortFacetValuesBy` setting.

Note for the documentation team (to be moved in the tracking issue): Using the new `sortFacetValuesBy` setting can slow down the facet-search requests as Meilisearch iterates over the whole list of facet values and computes the count of documents on every entry. That is hardly or even impossible to optimize correctly.

### TODO
 - [x] Create a custom HashMap wrapper for the facet `OrderBy` settings.
         This wrapper will return the `OrderBy` setting of the facet, if not defined will use the default `*` one, and if not there either (strange) will fall back on the lexicographic one.
- [x] Create a `ValuesCollection` wrapper that implements the logic for the lexicographic and count order by.
  - [x] Use it when there is no search query.
  - [x] Use it when there is a search query with and without allowed typos.
  - [x] Do not change the original logic, only use a wrapper.
- [x] Add tests

### CODE REVIEW

File: milli/src/order_by_map.rs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        self.0.into_iter()
+    }
+}
+
+impl<'de> Deserialize<'de> for OrderByMap {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let mut map = Deserialize::deserialize(deserializer).map(OrderByMap)?;
```
Comment (irevoire) [inline milli/src/order_by_map.rs:52]: Nice

File: milli/src/search/facet/search.rs [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+
+    pub fn query(&mut self, query: impl Into<String>) -> &mut Self {
+        self.query = Some(query.into());
+        self
+    }
+
+    pub fn max_values(&mut self, max: usize) -> &mut Self {
+        self.max_values = max;
+        self
+    }
```
Comment (irevoire) [inline milli/src/search/facet/search.rs:51]: Is there a true maximum in meilisearch? We can’t let users ask for as many values as they want, right?

File: milli/src/search/facet/search.rs [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+}
+
+impl ValuesCollection {
+    pub fn by_lexicographic(max: usize) -> Self {
+        ValuesCollection::Lexicographic { max, content: Vec::with_capacity(max) }
+    }
+
+    pub fn by_count(max: usize) -> Self {
+        ValuesCollection::Count { max, content: BinaryHeap::with_capacity(max) }
+    }
```
Comment (irevoire) [inline milli/src/search/facet/search.rs:282]: Here if the max hasn’t been checked prior, we could consume a ton of ram

**Review by irevoire — APPROVED** [review]

File: milli/src/search/facet/search.rs (hunk 3 above)
Comment (Kerollmops) [reply milli/src/search/facet/search.rs:282]: You are right. I will replace the `with_capacity` with `new` in this case. There is not important point in optimizing this small part.

File: milli/src/search/facet/search.rs (hunk 2 above)
Comment (Kerollmops) [reply milli/src/search/facet/search.rs:51]: There is no actual limit, but there is a warning in the documentation.

**Review by irevoire — APPROVED** [review]
bors merge

**Pull request merged**